> # CST1510 Programming for data communication and networks
> ## Week 2: Control Structures (Selection & Iteration) + Lists
> ## LAB ACTIVITY

<hr style="border:2px solid gray">


Welcome to your second Python lab session! This week, we'll focus on Python's **Conditional statements (if, elif, else)**, **loops (for, while)** and basic **list** operations.

By the end of this session, you'll be well-equipped to move forward in your Python journey.

### Topics Covered:

- **Learn to control the flow of your programs:** Understand how to make your code smarter by using conditional statements and loops.

- **Handle lists effectively:** Manage and manipulate data collections to create more efficient programs.


Once you’ve completed the tasks, select one or two activities demonstrating what you learned this week to include in your portfolio.

Remember, your portfolio should reflect your comprehension of variables, data types, and basic input/output operations—so choose wisely!
<hr style="border:1px solid gray">

# ACTIVITY:
<hr style="border:1px solid gray">

# Threshold:

1. **Password Strength Checker** 

**Scenario:**
You are a junior security analyst at a company. Employees often choose weak passwords, so you need to create a simple program to check password strength. Since you only know the basics of programming, your program will use simple rules.

**Requirements:**

1. Ask the user to enter a password.

2. If the password has fewer than 8 characters, print "Weak".

3. If the password has 8 or more characters, print "Moderate".

4. If the password has 12 or more characters, print "Strong"

In [6]:
password = input("Enter a password")

if len(password) < 8:
    print("Weak")
elif len(password) >= 12:
    print("Strong")
else:
    print("Moderate")



Enter a password bleedem01


Moderate


2. **Multiplication quiz**

Suppose you want to develop a program to help a first-grader practice multiplication. The program randomly generates two single-digit integers, number1 and number2, and displays to the student a question such as; What is 1 * 7. After the student types the answer, the program displays a message to indicate whether it is true or false.

To generate a random number, you can use the randint(a, b) function in the random module. This function returns a random integer i between a and b, inclusively. To obtain a random integer between 0 and 9, use randint(0, 9).

The program may be set up to work as follows:
Step 1: Randomly generate two single-digit integers for number1 (e.g., 4) and number 2 (e.g., 5)
Step 2: Prompt the student to answer, "What is 4 * 5?" (user input)
Step 3: Check whether the student’s answer is correct. 

In [12]:
import random
number1 = random.randint(0, 9)
number2 = random.randint(0, 9)

print(f"What is {number1} * {number2}?")
answer = int(input("Your answer: "))

if answer == number1 * number2:
    print("Answer is Correct")
else:
    print(f"Wrong...The correct answer is {number1 * number2}.")


What is 9 * 6?


Your answer:  54


Answer is Correct


3. **Even or Odd** 
Write a program that reads an integer from the user.  Then your program should display a message indicating whether the integer is even or odd. 


In [16]:
number=int(input("Enter a number"))

if number % 2 == 0:
    print("Number is Even")
else:
    print("Number is Odd")

Enter a number 5


Number is Odd


4. **Body mass index (BMI)** is a measure of health and weight.

It can be calculated by taking your weight in kilograms and dividing it by the square of your height in meters. The interpretation of BMI for people 20 years or older is as follows:


BMI	Interpretation
BMI < 18.5	Underweight
18.5 ≤ BMI ≤ 25.0	Normal
25.0 ≤ BMI ≤ 30.0	Overweight
30.0 ≤ BMI 	Obese
 
Write a program that prompts the user to enter weight in pounds and height in inches and display the BMI. Note that one pound is 0.45359237 kilograms, and one inch is 0.0254 meters


In [17]:
weight_lbs=float(input("Enter your weight in pounds"))
height_inches=float(input("Enter your height in inches"))

weight_kgs=weight_lbs * 0.45359237
height_m=height_inches * 0.0254

bmi = weight_kgs / (height_m ** 2)
print(f"BMI is: {bmi:.2f}")

if bmi < 18.5:
    print("Interpretation is Underweight")
elif bmi <= 25.0:
    print("Interpretation is Normal")
elif bmi <= 30.0:
    print("Interpretation is Overweight")
else:
    print("Interpretation is Obese")


Enter your weight in pounds 175
Enter your height in inches 6.0


BMI is: 3417.70
Interpretation is Obese


5. **Student Marks Analysis**
You are working as a data analyst in a school. You have the exam marks of students stored in a list. Your task is to analyze the data.

**Requirements**:

- Store marks of students in a list.

- Use a loop to calculate the average mark.

- Find the highest and lowest mark using list functions.

- Print the number of students who passed (marks ≥ 50) and failed.

In [18]:
marks = [56, 68, 22, 54, 41, 48, 81, 59, 93, 72]

total = 0
for mark in marks:
    total += mark
average = total / len(marks)

highest = max(marks)
lowest = min(marks)

passed = 0
failed = 0
for mark in marks:
    if mark >= 50:
        passed += 1
    else:
        failed += 1

print("Average mark is ", average)
print("Highest mark is ", highest)
print("Lowest mark is ", lowest)
print("Number of students passed is ", passed)
print("Number of students failed is ", failed)



Average mark is  59.4
Highest mark is  93
Lowest mark is  22
Number of students passed is  7
Number of students failed is  3


6 . Display numbers from `-10` to `-1` using for loop

```Expected output:

-10
-9
-8
-7
-6
-5
-4
-3
-2
-1
```

In [19]:

for number in range(-10, 0):
    print(number)


-10
-9
-8
-7
-6
-5
-4
-3
-2
-1



7. Write a program to print the following number pattern using a loop



```Expected output:
1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
```


In [20]:

rows = 5

for row_number in range(1, rows + 1):
    for number in range(1, row_number + 1):
        print(number, end=" ")
    print()


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 



8. Calculate the sum of all numbers from `1` to a `given number`

Write a program to accept a number `n` from a user and calculate the sum of all numbers from 1 to a given number n


For example, if the user entered 10 the output should be 55 (1+2+3+4+5+6+7+8+9+10)

```
Expected output
Enter the value n: 10
The sum of all numbers until 10 is 55
```




In [25]:
n=int(input("Enter a number"))
total=0

for number in range(1, n + 1):
    total += number
    
    print(f"The sum of all numbers until {n} is {total}")

Enter a number 10


The sum of all numbers until 10 is 1
The sum of all numbers until 10 is 3
The sum of all numbers until 10 is 6
The sum of all numbers until 10 is 10
The sum of all numbers until 10 is 15
The sum of all numbers until 10 is 21
The sum of all numbers until 10 is 28
The sum of all numbers until 10 is 36
The sum of all numbers until 10 is 45
The sum of all numbers until 10 is 55


9. Write a program to print multiplication table of a given number

For example, `num = 2`

```
Expected output:

0 x 2 = 0
1 x 2 = 2
2 x 2 = 4
3 x 2 = 6
4 x 2 = 8
5 x 2 = 10
6 x 2 = 12
7 x 2 = 14
8 x 2 = 16
9 x 2 = 18
```



In [32]:

num = int(input("Enter a number"))

for number in range(0, 10):
    product = number * num
    print(f"{number} x {num} = {product}")


Enter a number 2


0 x 2 = 0
1 x 2 = 2
2 x 2 = 4
3 x 2 = 6
4 x 2 = 8
5 x 2 = 10
6 x 2 = 12
7 x 2 = 14
8 x 2 = 16
9 x 2 = 18


10. Displays the following multiplication table. Consider the spacing between digits

```
Expected output:

0   0   0   0   0   0   0   0   0   0 
0   1   2   3   4   5   6   7   8   9 
0   2   4   6   8  10  12  14  16  18 
0   3   6   9  12  15  18  21  24  27 
0   4   8  12  16  20  24  28  32  36 
0   5  10  15  20  25  30  35  40  45 
0   6  12  18  24  30  36  42  48  54 
0   7  14  21  28  35  42  49  56  63 
0   8  16  24  32  40  48  56  64  72 
0   9  18  27  36  45  54  63  72  81 
```



In [37]:

for row in range(10):
    for column in range(10):
        print(f"{row * column:>3}", end=" ") #>3 means the spaces after each number
    #to move to the next row
    print()  


  0   0   0   0   0   0   0   0   0   0 
  0   1   2   3   4   5   6   7   8   9 
  0   2   4   6   8  10  12  14  16  18 
  0   3   6   9  12  15  18  21  24  27 
  0   4   8  12  16  20  24  28  32  36 
  0   5  10  15  20  25  30  35  40  45 
  0   6  12  18  24  30  36  42  48  54 
  0   7  14  21  28  35  42  49  56  63 
  0   8  16  24  32  40  48  56  64  72 
  0   9  18  27  36  45  54  63  72  81 


11. Use the math.sqrt function. Write a program that prints the following table using the sqrt function in the math module.

```
Number      Square root
0   	 	0.00
2    		1.41
4    		2.00
6    		2.45
8    		2.83
10    		3.16
12    		3.46
14    		3.74
16    		4.00
18    		4.24
```


In [38]:
import math

print("Number\tSquare root")

# 
for number in range(0, 20, 2):
    sqrt_value = math.sqrt(number)
    
    print(f"{number}\t{sqrt_value:.2f}")


Number	Square root
0	0.00
2	1.41
4	2.00
6	2.45
8	2.83
10	3.16
12	3.46
14	3.74
16	4.00
18	4.24


# Typical: 

12. **Dog Years**

It is commonly said that one human year is equivalent to 7 dog years. However, this simple conversion fails to recognise that dogs reach adulthood in approximately two years. As result, some people believe that it is better to count each of the first two human years as 10.5 dog years, and then count each additional human year as 4 dog years. 

Write a program that implements the conversion from human years to dog years described in the previous paragraph. Ensure that your program works correctly for conversions of less than two or more humans’ years. Your program should display an appropriate error message if the user enters a negative number. 


In [34]:
human_years = float(input("Enter the number of human years: "))

if human_years < 0:
    print("Error, Age shouldn't be negative.")
else:
    if human_years <= 2:
        dog_years = human_years * 10.5
    else:
        dog_years = 2 * 10.5 + (human_years - 2) * 4

    print(f"{human_years} human years is approximately {dog_years} dog years.")


Enter the number of human years:  20


20.0 human years is approximately 93.0 dog years.



13. **Vowel or Constant**

In this exercise you will create a program that read a letter of the alphabet from the user. If the user enter a, e, i, o or u then your program should display a message indicating that the entered letter is a vowel. If the user enters y then your program should display a message indicating that sometimes y is a vowel, and sometimes y is a consonants.  Otherwise your program should display a message indicating that the letter is a constant. 


In [6]:
letter = input("Enter a letter")
if letter in ["a","e","i","o","u"]:
    print("This letter is a vowel")
elif letter == "y":
    print("Sometimes y is a vowel,sometimes a consonant")
else:
    print("Letter is a consonant")

Enter a letter a


This letter is a vowel


14. Print the following pattern, write a program to use for loop to print the following reverse number pattern

```
Expected output:
5 4 3 2 1 
4 3 2 1 
3 2 1 
2 1 
1
```


In [15]:
for i in range(5,0,-1):
    for j in range(i,0,-1):
        print(j, end = ' ')
    print()

5 4 3 2 1 
4 3 2 1 
3 2 1 
2 1 
1 


15. **Find the highest score**

Write a program that prompts the user to enter the number of students and each student’s name and score and display the student name and highest score.


In [33]:
number = int(input("Enter the number of students: "))

scores = [int(input(f"Enter score {i+1}: ")) for i in range(number)]

scores.sort(reverse=True)

print("Highest score:", scores[0])
if number > 1:
    print("Second highest score:", scores[1])
else:
    print("There is no second highest score.")


Enter the number of students:  5
Enter score 1:  67
Enter score 2:  78
Enter score 3:  98
Enter score 4:  83
Enter score 5:  46


Highest score: 98
Second highest score: 83


16. Write a **guessing game** in Python where the user has to guess a secret number. After every guess the program tells the user whether their number was too large or too small. At the end the number of tries needed should be printed.


In [11]:
import random

secret_number = random.randint(1, 25)
tries = 0
guess = 0

print("Welcome to the Guessing Game")
print("Think of a number between 1 and 25.")

while guess != secret_number:
    guess = int(input("Enter your guess: "))
    tries += 1
if guess < secret_number:
    print("Number is too small.")
elif guess > secret_number:
    print("Number is too large.")
else:
    print(f"Correct,The secret number was {secret_number}.")
    print(f"You guessed in {tries} tries.")

Welcome to the Guessing Game
Think of a number between 1 and 25.


Enter your guess:  14
Enter your guess:  18
Enter your guess:  19
Enter your guess:  11
Enter your guess:  12
Enter your guess:  10
Enter your guess:  8
Enter your guess:  7
Enter your guess:  21
Enter your guess:  5
Enter your guess:  4
Enter your guess:  3
Enter your guess:  1
Enter your guess:  16
Enter your guess:  23
Enter your guess:  22
Enter your guess:  6
Enter your guess:  9


Correct,The secret number was 9.
You guessed in 18 tries.



17. **Collecting Positive Numbers**
You are writing a small program to help a teacher keep track of positive numbers entered by students. The teacher also wants to know if any even numbers were given.

**Requirements:**

1. Keep asking the user to enter a positive number, or type `"done"` to stop.
2. Store all entered numbers in a **list**.
3. After the input ends:

   * If numbers were entered:

     * Print all the numbers entered.
     * Check the list and print the **first even number**. If no even number exists, print `"No even numbers found"`.
   * If no numbers were entered, print `"No numbers were entered"`.


> ### Note 
>
>* **`lower()`**: Converts any text into lowercase.
>  Example: `"DONE".lower()` becomes `"done"`.
>  This makes it easier to compare user input without worrying about uppercase/lowercase.
>
>* **`isdigit()`**: Checks if a string only contains digits (0–9).
>  Example: `"123".isdigit()` returns `True`, but `"12a".isdigit()` returns `False`.
>  This helps us make sure the user typed a valid number before converting it into an integer.
> 


---

In [23]:
positive_numbers = []

while True:
    num = input("Enter a positive number: ")

    if num.lower() == "done":
        break

    if num.isdigit():
        num1 = int(num)
        if num1 > 0:
            positive_numbers.append(num1)
        else:
            print("Enter a positive number.")
    else:
        print("Please enter a number or 'done'.")

if len(positive_numbers) == 0:
    print("No numbers were entered.")
else:
    print("Numbers entered:", positive_numbers)

    first_even = None
    for n in positive_numbers:
        if n % 2 == 0:
            first_even = n
            break

    if first_even is not None:
        print("The first even number is:", first_even)
    else:
        print("No even numbers found.")


Enter a positive number:  5
Enter a positive number:  7
Enter a positive number:  8
Enter a positive number:  4
Enter a positive number:  3
Enter a positive number:  done


Numbers entered: [5, 7, 8, 4, 3]
The first even number is: 8


# Excellent:

18. Suppose you want to develop a **program to play a lottery.**
    
The program randomly generates a two-digit number, prompts the user to enter a <u>two-digit number</u>, and determines whether the user wins according to the following rules:


    1. If the user’s input matches the lottery in the exact order, the award is £10,000.
    2. If all the digits in the user’s input match all the digits in the lottery number, the award is £3,000.
    3. If one digit in the user’s input matches a digit in the lottery number, the award is £1,000.



In [24]:
import random

lottery = random.randint(1, 100)

guess = int(input("Enter a two digit number: "))
print("The lottery number is:", lottery)

if guess == lottery:
    print("You win £10,000!")

else:
    print("Guess is wrong!.")

Enter a two-digit number:  57


The lottery number is: 6
Guess is wrong!.



19. The program prompts the user to enter a <u>three-digit number</u> and determines whether the user wins according to the following rules:

    1. If the user input matches the lottery number in the exact order, the award is £10,000.
    2. If all the digits in the user input match all the digits in the lottery number, the award is £3,000.
    3. If one digit in the user input matches a digit in the lottery number, the award is £1,000.

In [27]:
import random

lottery = random.randint(100,999)
guess = int(input("Enter a three digit number: "))
print("The lottery number is: ", lottery)

if guess == lottery:
    print("You win £10,000!")

else:
    print("Guess is wrong!")

Enter a three digit number:  450


The lottery number is:  174
Guess is wrong!


20. (Find the two highest scores) - Write a program that prompts the user to enter the number of students and each student’s score and displays the highest and second highest scores.


In [32]:
students = int(input("Enter the number of students: "))
scores = []

for i in range(students):
    score = int(input(f"Enter score for student {i +1 }: "))
    scores.append(score)

if students == 1:
    print(f"Highest score: {scores[0]}")
    print("There is no second highest score.")
else:
    scores.sort(reverse=True)

    print(f"Highest score: {scores[0]}")
    print(f"Second highest score: {scores[1]}")

Enter the number of students:  6
Enter score for student 1:  87
Enter score for student 2:  80
Enter score for student 3:  94
Enter score for student 4:  76
Enter score for student 5:  62
Enter score for student 6:  79


Highest score: 94
Second highest score: 87



21. **Password cracker** - write a program to print the accuracy of password guess.

For example: The actual password is `MDX123` while a user tried `MDX321` or `mdx123`, the program should tell that how accurately the password was guessed, in this case it should be `50%` accuracy. if a user enter `MDX124`, the output should be `83%` and in case of `123MDX` the output should be `0%`.

Points to keep in mind while calculating the accuracy a) case of the letters entered b) order of the letters

If a password is accurately guessed, the output should state `Password cracked!`

---

## Portfolio
Select one or two completed activities that best demonstrate this week’s learning. Polish the code (naming, formatting, comments) so you can commit them to Git in later weeks
